tests/types/ember-concurrency-test.ts

```diff
@@ -88,12 +98,224 @@ module('unit tests', () => {
+    }
+
+    {
+      let d = { foo: 'foo', value: 123, *perform() { return 'foo' } };
+      expect<EncapsulatedTaskState<typeof d>>().toEqualTypeOf<{ foo: string }>();
+    }
+  });
+
+  test('TaskForEncapsulatedTaskDescriptor', () => {
+    {
+      let d = { foo: 'foo', *perform() {} };
+      expect<TaskForEncapsulatedTaskDescriptor<typeof d>>().toEqualTypeOf<EncapsulatedTask<void, [], { foo: string }>>();
+    }
+
+    {
+      let d = { foo: 'foo', *perform(foo: boolean) {} };
+      expect<TaskForEncapsulatedTaskDescriptor<typeof d>>().toEqualTypeOf<EncapsulatedTask<void, [boolean], { foo: string }>>();
+    }
+
+    {
+      let d = { foo: 'foo', *perform(foo: boolean, bar?: number) {} };
+      expect<TaskForEncapsulatedTaskDescriptor<typeof d>>().toEqualTypeOf<EncapsulatedTask<void, [boolean, number?], { foo: string }>>();
+    }
+
+    {
+      let d = { foo: 'foo', *perform() { return 'foo' } };
+      expect<TaskForEncapsulatedTaskDescriptor<typeof d>>().toEqualTypeOf<EncapsulatedTask<string, [], { foo: string }>>();
+    }
+
+    {
+      let d = { foo: 'foo', *perform(foo: boolean, bar?: number) { return 'foo' } };
+      expect<TaskForEncapsulatedTaskDescriptor<typeof d>>().toEqualTypeOf<EncapsulatedTask<string, [boolean, number?], { foo: string }>>();
+    }
+
+    {
+      let d = { foo: 'foo', perform() {} };
+      // @ts-expect-error
+      expect<TaskForEncapsulatedTaskDescriptor<typeof d>>();
+    }
+  });
+
+  test('TaskInstanceForEncapsulatedTaskDescriptor', () => {
+    {
+      let d = { foo: 'foo', *perform() {} };
+      expect<TaskInstanceForEncapsulatedTaskDescriptor<typeof d>>().toEqualTypeOf<TaskInstance<void> & { foo: string }>();
+    }
+
+    {
+      let d = { foo: 'foo', *perform(foo: boolean) {} };
+      expect<TaskInstanceForEncapsulatedTaskDescriptor<typeof d>>().toEqualTypeOf<TaskInstance<void> & { foo: string }>();
+    }
+
+    {
+      let d = { foo: 'foo', *perform(foo: boolean, bar?: number) {} };
+      expect<TaskInstanceForEncapsulatedTaskDescriptor<typeof d>>().toEqualTypeOf<TaskInstance<void> & { foo: string }>();
+    }
+
+    {
+      let d = { foo: 'foo', *perform() { return 'foo' } };
+      expect<TaskInstanceForEncapsulatedTaskDescriptor<typeof d>>().toEqualTypeOf<TaskInstance<string> & { foo: string }>();
+    }
+
+    {
+      let d = { foo: 'foo', *perform(foo: boolean, bar?: number) { return 'foo' } };
+      expect<TaskInstanceForEncapsulatedTaskDescriptor<typeof d>>().toEqualTypeOf<TaskInstance<string> & { foo: string }>();
+    }
+
+    {
+      let d = { foo: 'foo', perform() {} };
+      // @ts-expect-error
+      expect<TaskInstanceForEncapsulatedTaskDescriptor<typeof d>>();
+    }
+  });
+
   test('Task', () => {
     // @ts-expect-error
     new Task(); // Task cannot be constructed
 
     // @ts-expect-error
-    class Foo extends Task {} // Task cannot be subclassed
+    class Foo extends Task<any, any[]> {} // Task cannot be subclassed
```

Comment:
This is my only concern with the tweaks here as I see it, and I *think* it’s fine because inference works correctly in the cases we *don’t* expect to be errors. 🤔 We could in principle set the type args to default with `= any` and `= any[]`, but that might have other undesirable knock-on effects? In general, that’s the move I tend to reach for in scenarios like this where you want to preserve backwards compatibility while needing to make this kind of change. However, again, I think it’s needless here?

Reply:
This was not related to the PR. The genetics are already there and requires, this is trying to test that it errors because it cannot be subclasses, but was previously erroring because The generics are missing.

Reply:
Ah, very good—I thought it was odd that the changes you made would have required this. 😅